packages/@aws-cdk/aws-elasticsearch/lib/domain.ts

```diff
@@ -0,0 +1,1271 @@
+    });
+  }
+
+  /**
+   * Metric for JVM memory pressure.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricJVMMemoryPressure(props?: MetricOptions): Metric {
+    return this.metric('JVMMemoryPressure', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for master CPU utilization.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricMasterCPUUtilization(props?: MetricOptions): Metric {
+    return this.metric('MasterCPUUtilization', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for master JVM memory pressure.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricMasterJVMMemoryPressure(props?: MetricOptions): Metric {
+    return this.metric('MasterJVMMemoryPressure', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for KMS key errors.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricKMSKeyError(props?: MetricOptions): Metric {
+    return this.metric('KMSKeyError', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for KMS key being inaccessible.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricKMSKeyInaccessible(props?: MetricOptions): Metric {
+    return this.metric('KMSKeyInaccessible', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for number of searchable documents.
+   *
+   * @default maximum over 5 minutes
+   */
+  public metricSearchableDocuments(props?: MetricOptions): Metric {
+    return this.metric('SearchableDocuments', {
+      statistic: Statistic.MAXIMUM,
+      ...props,
+    });
+  }
+
+  /**
+   * Metric for search latency.
+   *
+   * @default p99 over 5 minutes
+   */
+  public metricSearchLatency(props?: MetricOptions): Metric {
+    return this.metric('SearchLatencyP99', { statistic: 'p99', ...props });
+  }
+
+  /**
+   * Metric for indexing latency.
+   *
+   * @default p99 over 5 minutes
+   */
+  public metricIndexingLatency(props?: MetricOptions): Metric {
+    return this.metric('IndexingLatencyP99', { statistic: 'p99', ...props });
+  }
+
+  private grant(
+    grantee: iam.IGrantable,
+    domainActions: string[],
+    resourceArn: string,
+    ...otherResourceArns: string[]
+  ): iam.Grant {
+    const resourceArns = [resourceArn, ...otherResourceArns];
+
+    const grant = iam.Grant.addToPrincipal({
+      grantee,
+      actions: domainActions,
+      resourceArns,
+      scope: this,
+    });
+
+    return grant;
+  }
+}
+
+
+/**
+ * Reference to an Elasticsearch domain.
+ */
+export interface DomainAttributes {
+  /**
+   * The ARN of the Elasticsearch domain.
+   */
+  readonly domainArn: string;
+
+  /**
+   * The domain name of the Elasticsearch domain.
+   */
+  readonly domainName: string;
+
+  /**
+   * The domain endpoint of the Elasticsearch domain.
+   */
+  readonly domainEndpoint: string;
```

Comment:
Seems strange that all of these are required. If an ARN is given, all others can be inferred, same as for endpoint - so i'm not sure i see the value in this. I think just adding a `fromDomainArn` instead of `fromDomainAttributes` would suffice.

Reply:
>I think just adding a fromDomainArn instead of fromDomainAttributes would suffice.
Correct me if I'm wrong, but I don't _think_ you can infer the endpoint given an ARN for a domain. The endpoint seems to have a random suffix?

If ☝️ is true, then would it make sense to only have `fromDomainEndpoint`?

Reply:
Or is it acceptable to have `fromDomainArn` with `domainEndpoint` not being available in that context?

Reply:
Could you please post examples for these 3 properties of a given domain?

Reply:
VPC-enabled domain:
- domainArn: `arn:aws:es:us-east-1:<account>:domain/domain-in-vpc`
- domainName: `domain-in-vpc`
- domainEndpoint: `https://vpc-domain-in-vpc-he7vrcns6axbzxvnfudbyzw6du.us-east-1.es.amazonaws.com`

Internet domain:
- domainArn: `arn:aws:es:us-east-1:<account>:domain/internet-facing`
- domainName: `internet-facing`
- domainEndpoint: `https://search-internet-facing-qnoe56y2af2qsmbgyfmhjia6mm.us-east-1.es.amazonaws.com`

Reply:
Thanks! Ok so it looks like `domainName` can always be inferred, and thats it basically. So just remove it from `fromDomainAttributes`?